plan/aggregation_push_down.go

```diff
@@ -0,0 +1,259 @@
+		}
+	}
+	return
+}
+
+func (a *aggPushDownSolver) splitAggFuncsAndGbyCols(agg *Aggregation, join *Join) (valid bool,
+	leftAggFuncs, rightAggFuncs []expression.AggregationFunction,
+	leftGbyCols, rightGbyCols []*expression.Column) {
+	valid, leftAggFuncs, rightAggFuncs = a.collectAggFuncs(agg, join)
+	if !valid {
+		return
+	}
+	leftGbyCols, rightGbyCols = a.collectGbyCols(agg, join)
+	return
+}
+
+// addGbyCol adds a column to gbyCols. If a group by column has existed, it will not be added repeatedly.
+func (a *aggPushDownSolver) addGbyCol(gbyCols []*expression.Column, cols ...*expression.Column) []*expression.Column {
+	for _, c := range cols {
+		duplicate := false
+		for _, gbyCol := range gbyCols {
+			if c.Equal(gbyCol) {
+				duplicate = true
+				break
+			}
+		}
+		if !duplicate {
+			gbyCols = append(gbyCols, c)
+		}
+	}
+	return gbyCols
+}
+
+// checkValidJoin checks if this join should be pushed across. Temporarily we only consider inner join.
+func (a *aggPushDownSolver) checkValidJoin(join *Join) bool {
+	// TODO: Support outer join.
+	return join.JoinType == InnerJoin
+}
+
+// decompose splits an aggregate function to two parts: a final mode function and a partial mode function. Currently
+// there are no differences between partial mode and complete mode, so we can confuse them.
+func (a *aggPushDownSolver) decompose(aggFunc expression.AggregationFunction, schema expression.Schema, id string) ([]expression.AggregationFunction, expression.Schema) {
+	// Result is a slice because avg should be decomposed to sum and count. Currently we don't process this case.
+	result := []expression.AggregationFunction{aggFunc.Clone()}
+	for _, aggFunc := range result {
+		schema = append(schema, &expression.Column{
+			ColName:  model.NewCIStr(fmt.Sprintf("join_agg_%d", len(schema))), // useless but for debug
+			FromID:   id,
+			Position: len(schema),
+			RetType:  aggFunc.GetType(),
+		})
+	}
+	aggFunc.SetArgs(expression.Schema2Exprs(schema[len(schema)-len(result):]))
+	aggFunc.SetMode(expression.FinalMode)
+	return result, schema
+}
+
+func (a *aggPushDownSolver) allFirstRow(aggFuncs []expression.AggregationFunction) bool {
+	for _, fun := range aggFuncs {
+		if fun.GetName() != ast.AggFuncFirstRow {
+			return false
+		}
+	}
+	return true
+}
+
+// tryToPushDownAgg tries to push down an aggregate function into a join path. If all aggFuncs are first row, we won't
+// process it temporarily. If not, We will add additional group by columns and first row functions. We make a new aggregation
+// operator.
+func (a *aggPushDownSolver) tryToPushDownAgg(aggFuncs []expression.AggregationFunction, gbyCols []*expression.Column, join *Join, childIdx int) LogicalPlan {
+	child := join.GetChildByIndex(childIdx).(LogicalPlan)
+	if a.allFirstRow(aggFuncs) {
+		return child
+	}
+	agg := &Aggregation{
+		GroupByItems:    expression.Schema2Exprs(gbyCols),
+		baseLogicalPlan: newBaseLogicalPlan(Agg, a.alloc),
+	}
+	child.SetParents(agg)
+	agg.SetChildren(child)
```

Comment:
If childIdx == 0, child is left child. So line211 means that left child set parents.
Line211 and line262 are not equal?